tfworker/backends/s3.py

```diff
@@ -302,3 +321,171 @@ def filter_keys(paginator, bucket_name, prefix="/", delimiter="/", start_after="
                     yield content["Key"]
         except TypeError:
             pass
+
+
+class S3Handler(BaseHandler):
+    """The S3Handler class is a handler for the s3 backend"""
+
+    actions = ["plan", "apply"]
+    required_vars = []
+    _is_ready = False
+
+    def __init__(self, authenticator):
+        try:
+            self.execution_functions = {
+                "plan": {
+                    "check": self._check_plan,
+                    "post": self._post_plan,
+                },
+                "apply": {
+                    "pre": self._pre_apply,
+                },
+            }
+
+            self._authenticator = authenticator
+            self._s3_client = self._authenticator.backend_session.client("s3")
+
+        except Exception as e:
+            raise HandlerError(f"Error initializing S3Handler: {e}")
+
+    def is_ready(self):
+        if not self._is_ready:
+            filename = str(uuid4().hex[:6].upper())
+            if self._s3_client.list_objects(
+                Bucket=self._authenticator.bucket,
+                Prefix=f"{self._authenticator.prefix}/{filename}",
+            ).get("Contents"):
+                raise HandlerError(
+                    f"Error initializing S3Handler, remote file already exists: {filename}"
+                )
+            try:
+                self._s3_client.upload_file(
+                    "/dev/null",
+                    self._authenticator.bucket,
+                    f"{self._authenticator.prefix}/{filename}",
+                )
+            except boto3.exceptions.S3UploadFailedError as e:
+                raise HandlerError(
+                    f"Error initializing S3Handler, could not create file: {e}"
+                )
+            try:
+                self._s3_client.delete_object(
+                    Bucket=self._authenticator.bucket,
+                    Key=f"{self._authenticator.prefix}/{filename}",
+                )
+            except boto3.exceptions.S3UploadFailedError as e:
+                raise HandlerError(
+                    f"Error initializing S3Handler, could not delete file: {e}"
+                )
+            self._is_ready = True
+        return self._is_ready
+
+    def execute(self, action, stage, **kwargs):
+        # save a copy of the planfile to the backend state bucket
+        if action in self.execution_functions.keys():
+            if stage in self.execution_functions[action].keys():
+                self.execution_functions[action][stage](**kwargs)
+        return None
+
+    def _check_plan(self, planfile: Path, definition: str, **kwargs):
+        """check_plan runs while the plan is being checked, it should fetch a file from the backend and store it in the local location"""
+        # ensure planfile does not exist or is zero bytes if it does
+        remotefile = f"{self._authenticator.prefix}/{definition}/{planfile.name}"
```

Comment:
I just thought of this, but is there any chance of someone being able to apply a plan that isn't ready. For instance, if I plan, but don't apply b/c it doesn't look right, how do we prevent someone from coming along and running apply before plan without thinking?

Reply:
That scenario is definitely possible. This is the scenario I was thinking of when I mentioned the notion of adding some sort of identifier to a worker run. The backend storage in the first iteration wasn't intended for humans to run, but rather to be executed as part of a job or automation run. The unique identifier is definitely an additional item that's needed.